### DISCUSSION TITLE
Save the image rendered by the OffscreenRendergraph channel

### DISCUSSION THREAD

**Discussion by WdsPjm** [vsg-dev/VulkanSceneGraph#1017] [Q&A] [answered]
I only want to save the image rendered by OffscreenRendergraph. How do I do this
I tried it by referring to vsgrendertotexture and vsgscreenshot, but it didn't work

    auto context = vsg::Context::create(window->getOrCreateDevice());

    // Framebuffer with attachments
    VkExtent2D targetExtent{512, 512};
    auto offscreenCamera = createCameraForScene(vsg_scene, targetExtent);
    auto colorImage = vsg::ImageInfo::create();
    auto depthImage = vsg::ImageInfo::create();
    auto rtt_RenderGraph = createOffscreenRendergraph(*context, targetExtent, *colorImage, *depthImage);

    auto rtt_view = vsg::View::create(offscreenCamera, vsg_scene);
    rtt_RenderGraph->addChild(rtt_view);

    // Planes geometry that uses the rendered scene as a texture map
    vsg::ref_ptr<vsg::Node> planes = createPlanes(colorImage);
    auto camera = createCameraForScene(planes, window->extent2D());
    auto main_RenderGraph = vsg::createRenderGraphForView(window, camera, planes);
    
    // add close handler to respond to the close window button and pressing escape
    viewer->addEventHandler(vsg::CloseHandler::create(viewer));
    nestedCommandGraph = true;
    viewer->addEventHandler(vsg::Trackball::create(camera));
    auto rtt_commandGraph = vsg::CommandGraph::create(window);
    
    if (nestedCommandGraph)
    {
        std::cout<<"Nested CommandGraph, with nested RTT CommandGraph as a child on the main CommandGraph. "<<std::endl;
      //  rtt_commandGraph = vsg::CommandGraph::create(window);
        rtt_commandGraph->submitOrder = -1; // render before the main_commandGraph
        rtt_commandGraph->addChild(rtt_RenderGraph);

        auto main_commandGraph = vsg::CommandGraph::create(window);
        main_commandGraph->addChild(main_RenderGraph);
       // main_commandGraph->addChild(rtt_commandGraph); // rtt_commandGraph nested within main CommandGraph

        viewer->assignRecordAndSubmitTaskAndPresentation({rtt_commandGraph});
    }
    else if (separateCommandGraph)
    {
        std::cout<<"Seperate CommandGraph with RTT CommandGraph first, then main CommandGraph second."<<std::endl;
        auto rtt_commandGraph = vsg::CommandGraph::create(window);
        rtt_commandGraph->submitOrder = -1; // render before the main_commandGraph
        rtt_commandGraph->addChild(rtt_RenderGraph);

        auto main_commandGraph = vsg::CommandGraph::create(window);
        main_commandGraph->addChild(main_RenderGraph);

        viewer->assignRecordAndSubmitTaskAndPresentation({rtt_commandGraph, main_commandGraph});
    }
    else
    {
        std::cout<<"Single CommandGraph containing by the RTT and main RenderGraphs"<<std::endl;
        // Place the offscreen RenderGraph before the plane geometry RenderGraph
        auto commandGraph = vsg::CommandGraph::create(window);
        commandGraph->addChild(rtt_RenderGraph);
        commandGraph->addChild(main_RenderGraph);

        viewer->assignRecordAndSubmitTaskAndPresentation({commandGraph});
    }

    auto event = vsg::Event::create_if(use_vkEvent, window->getOrCreateDevice()); // Vulkan creates VkEvent in an unsignalled state
    // Add ScreenshotHandler to respond to keyboard and mouse events.
    auto screenshotHandler = ScreenshotHandler::create(event, colorFilename, depthFilename, options);
    viewer->addEventHandler(screenshotHandler);
    if (event) rtt_commandGraph->addChild(vsg::SetEvent::create(event, VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT));

    viewer->compile();

    if (multiThreading)
    {
        std::cout<<"Enabled multi-threading"<<std::endl;
        viewer->setupThreading();
    }

    // rendering main loop
    while (viewer->advanceToNextFrame())
    {
        // pass any events into EventHandlers assigned to the Viewer
        viewer->handleEvents();
        viewer->update();
        viewer->recordAndSubmit();
        if (screenshotHandler->do_image_capture) screenshotHandler->screenshot_image(rtt_commandGraph->window);
        
        viewer->present();
    }

**Comment by WdsPjm** [comment, marked as answer]
Just make the following changes in screenshot_image to save only images rendered by offscreenrender
   

  auto physicalDevice = window->getPhysicalDevice();
      //  auto swapchain = window->getSwapchain();

        // get the colour buffer image of the previous rendered frame as the current frame hasn't been rendered yet.  The 1 in window->imageIndex(1) means image from 1 frame ago.
       
        auto sourceImage = offscreenImageInfo->imageView->image; //  window->imageView(window->imageIndex(1))->image;

        VkFormat sourceImageFormat = offscreenImageInfo->imageView->image->format; // swapchain->getImageFormat();
        VkFormat targetImageFormat = sourceImageFormat;

**Reply by robertosfield** [reply]
My brain is befuddled due to a cold so not a in place to go figure out stuff.  Does the above reply mean you've solved the problem?

**Reply by WdsPjm** [reply]
I am very sorry to have wasted your valuable time. Yeah, I already took care of it.

**Reply by robertosfield** [reply]
I'll make your above reply as a answer then, you should be able to do this yourself too if you ever answer your own question :-)

**Reply by WdsPjm** [reply]
I'm sorry to hear that you have a cold. Wishing you a speedy recovery
